logservice.go

```diff
@@ -2,140 +2,177 @@ package main
-		Bucket:        aws.String("logs.screwdriver.cd"),
-		Key:           aws.String(path),
-		Body:          data,
-		ContentLength: aws.Int64(size),
-		ContentType:   aws.String(fileType),
+			// Process the logs for this step in the background
+			wg.Add(1)
+			go func(buildID, step string, c chan *logLine) {
+				defer wg.Done()
+				processLogs(uploader, buildID, step, c)
+			}(buildID, newLog.Step, c)
+		}
+
+		c <- newLog
 	}
 
-	if _, err := svc.PutObject(params); err != nil {
-		log.Fatalf("bad response: %s", err)
+	// Close any channel that might be open still
+	for _, v := range steps {
+		close(v)
 	}
+
+	wg.Wait()
+
+	return nil
 }
 
-func processLogs(logChan chan *logLine, buildID string) chan struct{} {
-	exitChan := make(chan struct{})
-	go func() {
-		logs := map[string][][]byte{}
-
-		t := time.Tick(2 * time.Second)
-		for {
-			select {
-			case line, ok := <-logChan:
-				if !ok {
-					for step, lines := range logs {
-						log.Println("Uploading...")
-						uploadLogToS3(lines, buildID, step)
-					}
-					exitChan <- struct{}{}
-					return
-				}
-
-				s3Line := s3LogLine{
-					Time:    line.Time,
-					Message: line.Message,
-					Line:    len(logs[line.Step]),
-				}
-				j, _ := json.Marshal(s3Line)
-				logs[line.Step] = append(logs[line.Step], j)
-			case <-t:
-				for step, lines := range logs {
-					log.Println("Uploading...")
-					uploadLogToS3(lines, buildID, step)
-				}
-			}
+type logLine struct {
+	Time    int64  `json:"t"`
+	Message string `json:"m"`
+	Step    string `json:"s"`
+}
 
-		}
-	}()
-	return exitChan
+type s3LogLine struct {
+	Time    int64  `json:"t"`
+	Message string `json:"m"`
+	Line    int    `json:"n"`
 }
 
-func scanFile(path, buildID string) error {
-	log.Println("Starting scan")
-	logChan := make(chan *logLine, 50)
-	exitChan := processLogs(logChan, buildID)
+func processLogs(u s3fileuploader.S3FileUploader, buildID, stepName string, logs chan *logLine) error {
+	log.Println("Starting step processing for", stepName)
 
-	file, err := os.Open(path)
+	cache, err := ioutil.TempFile("", stepName)
 	if err != nil {
 		return err
 	}
+	defer os.Remove(cache.Name())
+	encoder := json.NewEncoder(cache)
 
-	scanner := bufio.NewScanner(file)
-	for scanner.Scan() {
-		newLine := &logLine{}
-		text := scanner.Text()
-		if err := json.Unmarshal([]byte(text), newLine); err != nil {
-			return fmt.Errorf("Unable to unmarshal log line %q: %v", text, err)
-		}
+	s3key := path.Join(buildID, stepName)
 
-		logChan <- newLine
+	// Every upload is identical
+	lineCount := 0
+	savedLineCount := 0
+	upload := func() {
+		if lineCount == savedLineCount {
+			return
+		}
+		log.Println("Uploading", cache.Name())
+		u.Upload(bucket, s3key, cache)
+		savedLineCount = lineCount
 	}
 
-	close(logChan)
-	<-exitChan
-	return nil
-}
+	// Ship the logs every uploadInterval.
+	ticker := time.NewTicker(2 * time.Second)
+	defer ticker.Stop()
+	t := ticker.C
+
+	for {
+		select {
+		case l, ok := <-logs:
+			if !ok {
```

Comment:
Line 98 should clean up any log channel that is left open. I can think of 2 reasons the channels would still not close:

1) The FIFO never closes. In this case, we will sit forever at the `for scanner.Scan() {` line. The build process is probably stuck in this case, and we need build timeouts to resolve the situation.
2) There is a panic occurring. In this case, the program is going to crash anyway and we kind of don't care. We could close all remaining channels in a defer to tidy stuff up.